### PR TITLE
RFC: numerics reform

### PR DESCRIPTION
**Pull request by aturon**
This RFC is preparation for API stabilization for the `std::num` module.  The goal is to finish the simplification efforts started in [@bjz's reversal of the numerics hierarcy](https://github.com/rust-lang/rust/issues/10387).

Broadly, the proposal is to collapse the remaining numeric hierarchy in `std::num`, and to provide only limited support for generic programming (roughly, only over primitive numeric types that vary based on size).  Traits giving detailed numeric hierarchy can and should be provided separately through the Cargo ecosystem.

Thus, this RFC proposes to flatten or remove most of the traits currently provided by `std::num`, and generally to simplify the module as much as possible in preparation for API stabilization.

[Rendered](https://github.com/aturon/rfcs/blob/num-reform/active/0000-num-reform.md)

### CODE REVIEW

**Comment by andydude** [comment]
The only code associated with this comment are the traits below:

```rust
pub trait RotateLeft<RHS> {
    type Output;

    fn rotate_left(self, rhs: RHS) -> Self::Output;
}

pub trait RotateRight<RHS> {
    type Output;

    fn rotate_right(self, rhs: RHS) -> Self::Output;
}

pub trait SwapBytes<RHS> {
    fn swap_bytes(self) -> Self;

    fn from_be(x: Self) -> Self;
    fn from_le(x: Self) -> Self;
    fn to_be(self) -> Self;
    fn to_le(self) -> Self;
}
```

I would like to bring attention to the particular use case of using these subsets of the current `Int` trait implemented for `std::simd` types (like `i16x8`, `i32x4`, `i64x2`, `u16x8`, `u32x4`, `u64x2`). The reason why you can't implement `Int` for SIMD types is that even though `NumCast` is easy to implement as the broadcast operation (a SIMD type composed of 8 copies of the same value), the reverse operations in `ToPrimitive` cannot be implemented in a satisfactory way (You could take the first element, or the last element, but neither of these are attractive options). So there is no natural implementation of `ToPrimitive` for SIMD types.

A second reason why these method subsets are useful to separate from the `Int` trait is that the current `rotate_left` and `rotate_right` methods strictly enforce `RHS=usize`, which is typically how they are used, but there are algorithms that require these methods where `RHS=Self` (meaning each SIMD component is rotated a different amount). This is already expressible in the current `Shl` and `Shr` traits, which are intimately connected to the rotation methods. Imaging `Self` below replaced with `u32x4` in the following example specializations:

```
impl Shl<usize, Output=Self> for ... { ... }
impl Shl<Self, Output=Self> for ... { ... }
impl Shr<usize, Output=Self> for ... { ... }
impl Shr<Self, Output=Self> for ... { ... }
impl RotateLeft<usize, Output=Self> for ... { ... }
impl RotateLeft<Self, Output=Self> for ... { ... }
impl RotateRight<usize, Output=Self> for ... { ... }
impl RotateRight<Self, Output=Self> for ... { ... }
```

After some experimentation, it seems that `Shl<Self, Self>`, `Shr<Self, Self>`, `RotateLeft<Self, Self>`, and `RotateRight<Self, Self>` are already implemented for all SIMD types. It would be a great benefit for SIMD users if the other, more common shift and rotation methods were also built-in.

A side note (and perhaps outside the scope of this comment) is that `RotateLeft` and `RotateRight` could be renamed `Rol` and `Ror` as they can be implemented in single instructions on some architectures, even though they're not traditional C operations. Evidence can be found on [the LLVM mailing list](http://lists.cs.uiuc.edu/pipermail/llvmdev/2009-February/019943.html "rol/ror llvm instruction set") that others who have also considered them core operations.

In conclusion, there are at least 2 compelling reasons (the `ToPrimitive` reason and the `RotateLeft` reason) to separate out rotation operations from the `Int` trait so that SIMD types may benefit from sharing the same methods as primitive types.
